pkg/providers/v2/instances_test.go

```diff
@@ -0,0 +1,466 @@
+			providerID: "aws:///us-west-1a/i-1",
+		},
+		{
+			name:       "get instance (without private IP/DNS) provider ID",
+			instance:   makeInstance(2, "", "1.2.3.4", "", "instance-same.ec2.external", "running"),
+			providerID: "aws:///us-west-1a/i-2",
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			providerID, err := getInstanceProviderID(testCase.instance)
+			assert.NoError(t, err)
+			assert.Equal(t, testCase.providerID, providerID)
+		})
+	}
+}
+
+func TestInstanceExists(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockEC2 := mocks.NewMockEC2(mockCtrl)
+
+	fakeInstances := &instances{
+		ec2: mockEC2,
+	}
+
+	nodeName := "ip-192-168-0-1.ec2.internal"
+
+	tests := []struct {
+		name              string
+		node              *v1.Node
+		expectedEc2Output *ec2.DescribeInstancesOutput
+		expectedResult    bool
+	}{
+		{
+			name: "test InstanceExists with running instance",
+			node: makeNode(nodeName),
+			expectedEc2Output: &ec2.DescribeInstancesOutput{
+				Reservations: []*ec2.Reservation{
+					{
+						Instances: []*ec2.Instance{
+							makeInstance(0, "192.168.0.1", "1.2.3.4", "instance-same.ec2.internal", "instance-same.ec2.external", "running"),
+						},
+					},
+				},
+			},
+			expectedResult: true,
+		},
+		{
+			name: "test InstanceExists with stopping instance",
+			node: makeNode(nodeName),
+			expectedEc2Output: &ec2.DescribeInstancesOutput{
+				Reservations: []*ec2.Reservation{
+					{
+						Instances: []*ec2.Instance{
+							makeInstance(1, "192.168.0.1", "1.2.3.4", "instance-same.ec2.internal", "instance-same.ec2.external", "stopping"),
+						},
+					},
+				},
+			},
+			expectedResult: true,
+		},
+		{
+			name: "test InstanceExists with terminated instance (node without providerID)",
+			node: makeNodeWithoutProviderID(nodeName),
+			expectedEc2Output: &ec2.DescribeInstancesOutput{
+				Reservations: []*ec2.Reservation{
+					{
+						Instances: []*ec2.Instance{},
+					},
+				},
+			},
+			expectedResult: false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			mockEC2.EXPECT().DescribeInstances(gomock.Any()).Return(test.expectedEc2Output, nil)
+
+			exists, err := fakeInstances.InstanceExists(context.TODO(), test.node)
+
+			if err != nil {
+				t.Errorf("InstanceExists failed with node %v: %v", nodeName, err)
+			}
+
+			if exists != test.expectedResult {
+				t.Errorf("unexpected result, InstanceExists should return %v", test.expectedResult)
+			}
+		})
+	}
+}
+
+func TestInstanceShutdown(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockEC2 := mocks.NewMockEC2(mockCtrl)
+
+	fakeInstances := &instances{
+		ec2: mockEC2,
+	}
+
+	nodeName := "ip-192-168-0-1.ec2.internal"
+
+	tests := []struct {
+		name              string
+		node              *v1.Node
+		expectedEc2Output *ec2.DescribeInstancesOutput
```

Comment:
I would call this field `mockedEC2Output` instead, since we are not actually validating against this as an "expected" valuie.

Reply:
same applies for other tests

Reply:
gotcha, will update!